gh-135640: Adds type checking to ElementTree.ElementTree constructor

abstractedfox opened the pull request:
Addresses issue #135640 by adding type checking to the constructor and to `ElementTree._setroot`, since both have the opportunity to assign _root. In addition to addressing the primary issue (possible data loss), this makes relevant tracebacks more descriptive. 

It looks like there had once been asserts that did these checks. I couldn't find rationale for why they were commented out in blame (it looks like this was done on svn, and I'm not sure where to look to find those commit messages). I opted to use `raise` instead as I felt it would be more descriptive.

At present, the way `iselement` is written still allows an object of the wrong type to pass through if it has a `tag` attribute. Commit a72a98f shows there were once comments indicating that the current implementation was done for a reason, but it isn't clear to me why that is. I left it alone for now, but changing it to this:
```
def iselement(element):
    return isinstance(element, Element)
```
...also passed tests. If there's no disagreement, I think it would be good to change that too.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:
V
```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-135640 -->
* Issue: gh-135640
<!-- /gh-issue-number -->

abstractedfox committed:
Tenuous first fix

Need to switch machines to test on Linux

abstractedfox committed:
Remove commented assertions

abstractedfox committed:
Add tests for ElementTree constructor

Add tests verifying the new type checking behavior

abstractedfox committed:
Add appropriate text to an exception

picnixz commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -527,7 +527,9 @@ class ElementTree:
     def __init__(self, element=None, file=None):
-        # assert element is None or iselement(element)
+        if element is not None and not iselement(element):
+            raise TypeError(f"element must be etree.Element, "
```
I would appreciate either writing xml.etree.Element or Element only. 

I also suspect that the isinstance() check was removed so that elements not inheriting from xml.Element still work, or elements that implement the XML interface without directly inheriting from it.

And the minimal element interface is "tag".

picnixz commented on Lib/test/test_xml_etree.py:
```diff
@@ -247,6 +247,13 @@ def check_element(element):
         self.assertRegex(repr(element), r"^<Element 't\xe4g' at 0x.*>$")
         element = ET.Element("tag", key="value")
 
+        # Verify type checking for ElementTree constructor
```
Can you make a separate test method, say "test_constructor" and put that test before test_interface? TiA

picnixz commented on Lib/test/test_xml_etree.py:
```diff
@@ -247,6 +247,13 @@ def check_element(element):
         element = ET.Element("tag", key="value")
 
+        # Verify type checking for ElementTree constructor
+
```
Add also a test where only an Element-like class is tested (only with a tag attribute, nothing else, not even inheriting from xml.etree.Element).

abstractedfox replied:
Hmm, I tried this in a boilerplate file:
```
class element impl:
    def __init__(self, tag):
        self.tag = tag
```

..but an instance of that wasn't enough to be able to pass as a root element to an ElementTree instance, call `write`, and have it work as normal (assuming I understood you properly). I ended up needing to add `iter()`, `items()`, `__len__()`, `text`, and `tail` before it would. Was that what you meant?

As a side thought, if ElementTree supports custom element implementations, should I remove the type check and change `iselement` to check for all of those attributes instead?

picnixz reviewed (commented):
I think the tests are expecting more attributes than the runtime for an "Element": compare check_element() and iselement().

abstractedfox committed:
Make iselement stricter + incorporate feedback

Adjusts iselement() based off feedback from picnixz. Since ElementTree
needs to be compatible with element-like objects, the function now checks
for all attributes that are necessary to instantiate an ElementTree object
and successfully write it to a file (preventing issue #135640).
This commit also incorporates other feedback requesting moving the
tests and adding more tests, and adds a news entry.

vstinner commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -120,8 +120,9 @@ class ParseError(SyntaxError):
-
+    return (hasattr(element, 'tag') and hasattr(element, 'text') and
+        hasattr(element, 'tail') and callable(element.iter) and
+        callable(element.items) and callable(element.__len__))
```
Does this function raise an exception if iter(), items() or `__len__()` method doesn't exist?

vstinner commented on Lib/test/test_xml_etree.py:
```diff
@@ -218,6 +218,42 @@ class ElementTreeTest(unittest.TestCase):
+        try:
+            tree = ET.ElementTree(element_like)
+        except Exception as err:
+            self.fail(err)
```
Maybe test also _setroot() here?

abstractedfox committed:
Revert iselement to previous state

abstractedfox committed:
Add iselement check to ElementTree.write

abstractedfox committed:
Make errors more descriptive

efimov-mikhail commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -122,7 +122,6 @@ def iselement(element):
     """Return True if *element* appears to be an Element."""
     return hasattr(element, 'tag')
 
```
```suggestion


```

efimov-mikhail commented on Misc/NEWS.d/next/Library/2025-06-22-02-16-17.gh-issue-135640.FXyFL6.rst:
```diff
@@ -0,0 +1,3 @@
+Address bug where calling :func:`xml.etree.ElementTree.ElementTree.write` on
+an ElementTree object with an invalid root element would blank the file
+passed to ``write`` if it already existed.
```
Maybe rewrite this in more simple manner? Split one sentence into two, for example.

efimov-mikhail commented on Lib/test/test_xml_etree.py:
```diff
@@ -218,6 +218,42 @@ class ElementTreeTest(unittest.TestCase):
+                self.tag = "tag"
+                self.text = None
+                self.tail = None
+            def iter(self):
```
It's `iter`, not `__iter__`?

abstractedfox replied:
For the sake of testing what .write() needs to complete successfully, it is (used in `_namespaces` in ElementTree.py). But actually, that test should be changed anyway since it was made in anticipation of the changes to `iselement` that have been undone

abstractedfox committed:
Fix whitespace

Co-authored-by: Mikhail Efimov <efimov.mikhail@gmail.com>

abstractedfox committed:
Rephrase news entry

abstractedfox committed:
Test: remove unnecessary fields for element-like

Test was written for a change to `iselement` that has since been undone

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -527,7 +527,10 @@ class ElementTree:
     def __init__(self, element=None, file=None):
-        # assert element is None or iselement(element)
+        if element is not None and not iselement(element):
+            raise TypeError(f"element must be xml.etree.Element or "
```
It is `xml.etree.ElementTree.Element`.

Look at `_assert_is_element()`. We need a similar error message.

picnixz replied:
That's my bad, I forgot the `ElementTree` part in my suggestion

serhiy-storchaka reviewed (commented):
1. This does not completely prevent failure during writing. Even if you check for all attributes used by `write()`, as in your previous versions, you do not check that attributes has correct type, that methods are callable, that they have correct signature and semantic. You do not check that subelements has correct types. Checking all this would be expensive and will not guarantee that `write()` will not fail.
2. This is not a problem that requires expensive solution. If your program contains such bug, it will be caught in the first test run. The difference between AttributeError and TypeError is not so important. You don't run your program on production files without testing, right?

The only way to guarantee that the output file will not be corrupted due to errors is to write to temporary file or in-memory file, and then atomically replace the content.

abstractedfox committed:
Change error messages

vstinner reviewed (approved):
LGTM

efimov-mikhail commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -716,7 +716,8 @@ def write(self, file_or_filename,
         """
         if not iselement(self._root):
-            raise TypeError(f"Root element must be of type xml.etree.Element "
+            raise TypeError(f"Root element must be "
```
There is no need for f-strings.

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -709,6 +715,10 @@ def write(self, file_or_filename,
+        if not iselement(self._root):
+            raise TypeError(f"Root element must be "
+                            f"xml.etree.ElementTree.Element "
+                            f"or Element-like object")
```
```suggestion
        if self._root is None:
            raise TypeError('ElementTree not initialized')
```

serhiy-storchaka replied:
Please add also a test for writing uninitialized ElementTree:
```py
    def test_write_uninitialized(self):
        tree = ET.ElementTree()
        self.assertRaises(TypeError, tree.write, TESTFN)
        self.assertFalse(os.path.exists(TESTFN))
```

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -543,7 +546,10 @@ def _setroot(self, element):
+        if not iselement(element):
+            raise TypeError(f"expected an xml.etree.ElementTree.Element or "
+                            f"Element-like object, not "
+                            f"{type(element).__name__}")
```
```suggestion
            raise TypeError('expected an Element, not %s' % type(e).__name__)
```

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -527,7 +527,10 @@ class ElementTree:
+        if element is not None and not iselement(element):
+            raise TypeError(f"expected an xml.etree.ElementTree.Element or "
+                            f"Element-like object, not "
+                            f"{type(element).__name__}")
```
```suggestion
            raise TypeError('expected an Element, not %s' % type(e).__name__)
```

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -218,6 +218,34 @@ class ElementTreeTest(unittest.TestCase):
+        # Test _setroot as well, since it also sets the _root object.
+
+        tree = ET.ElementTree()
+        with self.assertRaises(TypeError):
+            tree._setroot("")
+        with self.assertRaises(TypeError):
+            tree._setroot(ET.ElementTree())
```
Test also `_setroot(None)`.

`_setroot()` needs a separate test. It should also be tested with valid argument. `getroot()` should be used to check its effect.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -218,6 +218,34 @@ class ElementTreeTest(unittest.TestCase):
+        try:
+            tree = ET.ElementTree(element_like)
+        except Exception as err:
+            self.fail(err)
```
This is not needed.

abstractedfox committed:
Correct error messages

Suggested changes from @serhiy-storchaka

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

abstractedfox committed:
Correct error message

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

abstractedfox committed:
Tests for _setroot

Split tests for _setroot into their own function and remove test for
element-like objects per feedback from @serhiy-storchaka

abstractedfox committed:
Fix names

serhiy-storchaka reviewed (approved):
LGTM.
